### PR TITLE
[core] Enable channelz and reflection for gRPC.

### LINKED ISSUE #30086: [Core] GCS hanging even cpu utilization low
### What happened + What you expected to happen

```

Thread 1 (Thread 0x7f04b613fb00 (LWP 1746)):
#0  __libc_sendmsg (flags=16384, msg=0x7ffea7f3e490, fd=147) at ../sysdeps/unix/sysv/linux/sendmsg.c:28
#1  __libc_sendmsg (fd=147, msg=0x7ffea7f3e490, flags=16384) at ../sysdeps/unix/sysv/linux/sendmsg.c:25
#2  0x00005633ab3266ae in tcp_send (fd=147, msg=0x7ffea7f3e490, additional_flags=0) at external/com_github_grpc_grpc/src/core/lib/iomgr/tcp_posix.cc:916
#3  0x00005633ab327305 in tcp_flush (tcp=0x7f046009a520, error=0x7ffea7f3f548) at external/com_github_grpc_grpc/src/core/lib/iomgr/tcp_posix.cc:1448
#4  0x00005633ab327ca7 in tcp_write (ep=0x7f046009a520, buf=0x7f04600883d0, cb=0x7f04600881b8, arg=0x0) at external/com_github_grpc_grpc/src/core/lib/iomgr/tcp_posix.cc:1588
#5  0x00005633ab33216b in grpc_endpoint_write (ep=0x7f046009a520, slices=0x7f04600883d0, cb=0x7f04600881b8, arg=0x0) at external/com_github_grpc_grpc/src/core/lib/iomgr/endpoint.cc:32
#6  0x00005633aafa7d83 in write_action (gt=0x7f0460088010) at external/com_github_grpc_grpc/src/core/ext/transport/chttp2/transport/chttp2_transport.cc:983
#7  0x00005633aafa7bd7 in write_action_begin_locked (gt=0x7f0460088010) at external/com_github_grpc_grpc/src/core/ext/transport/chttp2/transport/chttp2_transport.cc:957
#8  0x00005633ab36c408 in grpc_combiner_continue_exec_ctx () at external/com_github_grpc_grpc/src/core/lib/iomgr/combiner.cc:264
#9  0x00005633ab36e159 in grpc_core::ExecCtx::Flush (this=0x7ffea7f3f7a0) at external/com_github_grpc_grpc/src/core/lib/iomgr/exec_ctx.cc:82
#10 0x00005633aad5a033 in grpc_core::ExecCtx::~ExecCtx (this=0x7ffea7f3f7a0, __in_chrg=<optimized out>) at external/com_github_grpc_grpc/src/core/lib/iomgr/exec_ctx.h:117
#11 0x00005633ab2fd738 in grpc_call_start_batch (call=0x7f032c760ff0, ops=0x7ffea7f3f870, nops=3, tag=0x7f049902fd58, reserved=0x0) at external/com_github_grpc_grpc/src/core/lib/surface/call.cc:1767
#12 0x00005633aad93e7b in grpc::CoreCodegen::grpc_call_start_batch (this=0x5633ad4f0e90, call=0x7f032c760ff0, ops=0x7ffea7f3f870, nops=3, tag=0x7f049902fd58, reserved=0x0) at external/com_github_grpc_grpc/src/cpp/common/core_codegen.cc:113
#13 0x00005633aa6f4dad in grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallOpSendMessage, grpc::internal::CallOpServerSendStatus, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >::ContinueFil
lOpsAfterInterception (this=0x7f049902fd58) at external/com_github_grpc_grpc/include/grpcpp/impl/codegen/call_op_set.h:976
#14 0x00005633aa6f4b69 in grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallOpSendMessage, grpc::internal::CallOpServerSendStatus, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >::FillOps (this=0x7f049902fd58, call=0x7f049902fbc0) at external/com_github_grpc_grpc/include/grpcpp/impl/codegen/call_op_set.h:906
#15 0x00005633aad68c1c in grpc::Server::PerformOpsOnCall (this=0x5633ad54a770, ops=0x7f049902fd58, call=0x7f049902fbc0) at external/com_github_grpc_grpc/src/cpp/server/server_cc.cc:1305
#16 0x00005633aa61dc7b in grpc::internal::Call::PerformOps (this=0x7f049902fbc0, ops=0x7f049902fd58) at external/com_github_grpc_grpc/include/grpcpp/impl/codegen/call.h:68
#17 0x00005633aa710d18 in grpc::ServerAsyncResponseWriter<ray::rpc::GcsSubscriberPollReply>::Finish (this=0x7f049902fbb8, msg=..., status=..., tag=0x7f049902f800) at external/com_github_grpc_grpc/include/grpcpp/impl/codegen/async_unary_call.h:358
#18 0x00005633aa702b74 in ray::rpc::ServerCallImpl<ray::rpc::InternalPubSubGcsServiceHandler, ray::rpc::GcsSubscriberPollRequest, ray::rpc::GcsSubscriberPollReply>::SendReply (this=0x7f049902f800, status=...) at bazel-out/k8-dbg/bin/_virtual_includes/grpc_common_lib/ray
/rpc/server_call.h:247
#19 0x00005633aa7028ea in ray::rpc::ServerCallImpl<ray::rpc::InternalPubSubGcsServiceHandler, ray::rpc::GcsSubscriberPollRequest, ray::rpc::GcsSubscriberPollReply>::HandleRequestImpl()::{lambda(ray::Status, std::function<void ()>, std::function<void ()>)#1}::operator()(
ray::Status, std::function<void ()>, std::function<void ()>) const (this=0x7f049902f800, status=..., success=..., failure=...) at bazel-out/k8-dbg/bin/_virtual_includes/grpc_common_lib/ray/rpc/server_call.h:207
#20 0x00005633aa735319 in std::__invoke_impl<void, ray::rpc::ServerCallImpl<ray::rpc::InternalPubSubGcsServiceHandler, ray::rpc::GcsSubscriberPollRequest, ray::rpc::GcsSubscriberPollReply>::HandleRequestImpl()::{lambda(ray::Status, std::function<void ()>, std::function<
void ()>)#1}&, ray::Status, std::function<void ()>, std::function<void ()> >(std::__invoke_other, ray::rpc::ServerCallImpl<ray::rpc::InternalPubSubGcsServiceHandler, ray::rpc::GcsSubscriberPollRequest, ray::rpc::GcsSubscriberPollReply>::HandleRequestImpl()::{lambda(ray::Status, std::function<void ()>, std::function<void ()>)#1}&, ray::Status&&, std::function<void ()>&&, std::function<void ()>&&) (__f=...) at /opt/rh/devtoolset-10/root/usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:60
#21 0x00005633aa72e86b in std::__invoke_r<void, ray::rpc::ServerCallImpl<ray::rpc::InternalPubSubGcsServiceHandler, ray::rpc::GcsSubscriberPollRequest, ray::rpc::GcsSubscriberPollReply>::HandleRequestImpl()::{lambda(ray::Status, std::function<void ()>, std::function<void ()>)#1}&, ray::Status, std::function<void ()>, std::function<void ()> >(void&&, (ray::rpc::ServerCallImpl<ray::rpc::InternalPubSubGcsServiceHandler, ray::rpc::GcsSubscriberPollRequest, ray::rpc::GcsSubscriberPollReply>::HandleRequestImpl()::{lambda(ray::Status, std::f
unction<void ()>, std::function<void ()>)#1}&)...) (__fn=...) at /opt/rh/devtoolset-10/root/usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:110
#22 0x00005633aa720182 in std::_Function_handler<void (ray::Status, std::function<void ()>, std::function<void ()>), ray::rpc::ServerCallImpl<ray::rpc::InternalPubSubGcsServiceHandler, ray::rpc::GcsSubscriberPollRequest, ray::rpc::GcsSubscriberPollReply>::HandleRequestI
mpl()::{lambda(ray::Status, std::function<void ()>, std::function<void ()>)#1}>::_M_invoke(std::_Any_data const&, ray::Status&&, std::function<void ()>&&, ray::Status&&) (__functor=..., __args#0=..., __args#1=..., __args#2=...) at /opt/rh/devtoolset-10/root/usr/lib/gcc/
x86_64-redhat-linux/10/../../../../include/c++/10/bits/std_function.h:291
#23 0x00005633aa74c96a in std::function<void (ray::Status, std::function<void ()>, std::function<void ()>)>::operator()(ray::Status, std::function<void ()>, std::function<void ()>) const (this=0x7f04783e33f8, __args#0=..., __args#1=..., __args#2=...) at /opt/rh/devtools
et-10/root/usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/std_function.h:622
#24 0x00005633aa75299f in operator()(ray::Status, std::function<void()>, std::function<void()>) const (__closure=0x7f04783e33f0, status=..., success_cb=..., failure_cb=...) at src/ray/gcs/gcs_server/pubsub_handler.cc:77
#25 0x00005633aa7539f3 in std::__invoke_impl<void, ray::gcs::InternalPubSubHandler::HandleGcsSubscriberPoll(ray::rpc::GcsSubscriberPollRequest, ray::rpc::GcsSubscriberPollReply*, ray::rpc::SendReplyCallback)::<lambda(ray::Status, std::function<void()>, std::function<voi
d()>)>&, ray::Status, std::function<void()>, std::function<void()> >(std::__invoke_other, struct {...} &) (__f=...) at /opt/rh/devtoolset-10/root/usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:60
#26 0x00005633aa7537cc in std::__invoke_r<void, ray::gcs::InternalPubSubHandler::HandleGcsSubscriberPoll(ray::rpc::GcsSubscriberPollRequest, ray::rpc::GcsSubscriberPollReply*, ray::rpc::SendReplyCallback)::<lambda(ray::Status, std::function<void()>, std::function<void()
>)>&, ray::Status, std::function<void()>, std::function<void()> >(struct {...} &) (__fn=...) at /opt/rh/devtoolset-10/root/usr/lib/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/invoke.h:110
#27 0x00005633aa7535a6 in std::_Function_handler<void(ray::Status, std::function<void()>, std::function<void()>), ray::gcs::InternalPubSubHandler::HandleGcsSubscriberPoll(ray::rpc::GcsSubscriberPollRequest, ray::rpc::GcsSubscriberPollReply*, ray::rpc::SendReplyCallback)
::<lambda(ray::Status, std::function<void()>, std::function<void()>)> >::_M_invoke(const std::_Any_data &, ray::Status &&, std::function<void()> &&, std::function<void()> &&) (__functor=..., __args#0=..., __args#1=..., __args#2=...) at /opt/rh/devtoolset-10/root/usr/lib
/gcc/x86_64-redhat-linux/10/../../../../include/c++/10/bits/std_function.h:291
```

### Versions / Dependencies

master

### Reproduction script

create a lot of nodes and run

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
This PR enable channelz and reflection for gRPC.

Channelz is useful to debug grpc related issues (https://github.com/grpc/proposal/blob/master/A14-channelz.md)

gRPC reflection enable [grpcurl](https://github.com/fullstorydev/grpcurl) and [grpcui](https://github.com/fullstorydev/grpcui) to interact with gRPC servers.
With reflection turned on, we can query the gRPC server easily without protobuf predefined.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
with grpcurl:

image

File: src/ray/rpc/grpc_server.cc [hunk 1]
```diff
@@ -42,6 +44,8 @@ GrpcServer::GrpcServer(std::string name,
   // Enable built in health check implemented by gRPC:
   //   https://github.com/grpc/grpc/blob/master/doc/health-checking.md
   grpc::EnableDefaultHealthCheckService(true);
+  grpc::reflection::InitProtoReflectionServerBuilderPlugin();
```
Comment (rkooo567) [inline src/ray/rpc/grpc_server.cc:47]: Can we make this optional? I will be surprised if this has no perf impact. Maybe we can allow to enable it using env var?

**Review by rkooo567 — COMMENTED** [review]
How do we access the UI? (how does it choose the port and how do we find the port number?)

**Comment by iycheng** [comment]
> How do we access the UI? (how does it choose the port and how do we find the port number?)

You should follow this https://github.com/fullstorydev/grpcui

It's very simple and easy to do. We probably can add something in the debugging in our docs for this.

File: src/ray/rpc/grpc_server.cc (hunk 1 above)
Comment (scv119) [reply src/ray/rpc/grpc_server.cc:47]: or do a quick micro benchmark?  actually i think it's unlikely to affect normal RPCs.

**Review by scv119 — APPROVED** [review]
let's attach microbenchmark result before merging.

**Comment by iycheng** [comment]
https://buildkite.com/ray-project/release-tests-pr/builds/20649#01845e8c-0bf2-4327-b25d-c2495c5ebcfe

New:
```


single_client_get_calls_Plasma_Store = [5579.829481115008, 65.62929190492056]
--
  | single_client_put_calls_Plasma_Store = [5601.421720223946, 106.615567802396]
  | multi_client_put_calls_Plasma_Store = [11225.648073400926, 98.37657625992156]
  | single_client_put_gigabytes = [19.892785384882878, 5.047044581001871]
  | single_client_tasks_and_get_batch = [11.011345680364723, 0.13948094965882857]
  | multi_client_put_gigabytes = [27.37878560995965, 0.8002392386909606]
  | single_client_get_object_containing_10k_refs = [12.615451516326997, 0.447614405548985]
  | single_client_tasks_sync = [1206.503570077645, 20.287508009021508]
  | single_client_tasks_async = [11016.015094873215, 196.694363912234]
  | multi_client_tasks_async = [32645.99049181028, 1971.5518539229429]
  | 1_1_actor_calls_sync = [2121.3314009143514, 16.588515924369755]
  | 1_1_actor_calls_async = [5585.503004222507, 66.0423540666254]
  | 1_1_actor_calls_concurrent = [4496.837428050913, 30.249605954862847]
  | 1_n_actor_calls_async = [11544.333668156169, 62.86227807418838]
  | n_n_actor_calls_async = [35166.417282478644, 1289.9239660167889]
  | n_n_actor_calls_with_arg_async = [2705.1791190236195, 9.230736467375054]
  | 1_1_async_actor_calls_sync = [1408.3924697966245, 12.346223774154446]
  | 1_1_async_actor_calls_async = [3049.498772304734, 124.00085219368022]
  | 1_1_async_actor_calls_with_args_async = [2220.7142077567974, 82.16001235341189]
  | 1_n_async_actor_calls_async = [10522.021811817667, 216.32575952499573]
  | n_n_async_actor_calls_async = [27786.733263218008, 720.4226286151876]
  | placement_group_create/removal = [994.8047033825665, 9.347140985514912]
  | client__get_calls = [1034.7608053253796, 12.050028108488052]
  | client__put_calls = [734.8804786506095, 22.50950730330715]
  | client__put_gigabytes = [0.046324621869606335, 0.0003015874496292288]
  | client__tasks_and_put_batch = [10215.31079499008, 174.74496888973036]
  | client__1_1_actor_calls_sync = [459.4425266688643, 13.113764878537573]
  | client__1_1_actor_calls_async = [849.8661586709009, 18.95326269042512]
  | client__1_1_actor_calls_concurrent = [854.726383900062, 10.730955755279469]
  | client__tasks_and_get_batch = [0.7300295525097047, 0.022655609490912568]

```

vs

Old

```
single_client_get_calls_Plasma_Store = [5691.701914755293, 232.7949988738463]
  | single_client_put_calls_Plasma_Store = [5626.492554264843, 66.46090401668248]
  | multi_client_put_calls_Plasma_Store = [12103.027994982556, 104.89640880864754]
  | single_client_put_gigabytes = [20.02940214664503, 5.348719162707589]
  | single_client_tasks_and_get_batch = [11.13490658304157, 0.20861549565309748]
  | multi_client_put_gigabytes = [33.55499587782516, 1.6027266663741464]
  | single_client_get_object_containing_10k_refs = [12.446270720513338, 0.12299476381071435]
  | single_client_tasks_sync = [1210.05244114599, 7.088876153549394]
  | single_client_tasks_async = [11109.437574318508, 173.66286757804912]
  | multi_client_tasks_async = [31773.757419439193, 630.9680210272662]
  | 1_1_actor_calls_sync = [2093.889076306659, 21.46787283304899]
  | 1_1_actor_calls_async = [5697.89836164464, 238.6990862752297]
  | 1_1_actor_calls_concurrent = [4406.391562953829, 121.05008398355392]
  | 1_n_actor_calls_async = [11344.08746969225, 260.4176800061584]
  | n_n_actor_calls_async = [34171.83061620843, 848.2985529912291]
  | n_n_actor_calls_with_arg_async = [2767.6854129987637, 15.964308544714612]
  | 1_1_async_actor_calls_sync = [1445.200257374686, 33.37761440479422]
  | 1_1_async_actor_calls_async = [3006.92193709883, 40.991034778440714]
  | 1_1_async_actor_calls_with_args_async = [2083.086467122248, 65.43690962046138]
  | 1_n_async_actor_calls_async = [10645.731543911575, 64.81412709435669]
  | n_n_async_actor_calls_async = [26896.924647686665, 768.090822327973]
  | placement_group_create/removal = [1038.6990695434863, 9.021183010909294]
  | client__get_calls = [1015.5554232292307, 16.513133954093217]
  | client__put_calls = [758.1450489378345, 10.350247350513683]
  | client__put_gigabytes = [0.04603399844246967, 0.00041740138461736196]
  | client__tasks_and_put_batch = [10571.651563199572, 170.59693532471883]
  | client__1_1_actor_calls_sync = [416.7790551417661, 12.647993738369834]
  | client__1_1_actor_calls_async = [786.4866732281822, 26.22869392949586]
  | client__1_1_actor_calls_concurrent = [748.8151304584452, 10.883006702615614]
  | client__tasks_and_get_batch = [0.7095882066437538, 0.021453079147163406]
 

<br class="Apple-interchange-newline">
```

**Comment by iycheng** [comment]
seems ok to me. @scv119  to confirm.

**Review by rkooo567 — APPROVED** [review]
LGTM given there's no perf regression

**Comment by daturkel** [comment]
Sorry to comment on a very old thread, but does this feature still work? When I try querying a ray service with grpcurl now (without providing proto files), I get "server does not support the reflection API"

**Pull request merged**